File: src/internal/cache/cache.go
package cache

import (
	"crypto/sha256"
	"fmt"
	"io/ioutil"
	"net/http"
	"os"
	"path/filepath"
	"strings"
)

type Cache struct {
	BaseDir string
}

func New() (*Cache, error) {
	home, err := os.UserHomeDir()
	if err != nil {
		return nil, err
	}
	base := filepath.Join(home, ".zenith", "cache")
	if err := os.MkdirAll(base, 0755); err != nil {
		return nil, err
	}
	return &Cache{BaseDir: base}, nil
}

func (c *Cache) Get(url string) (string, error) {
	if !strings.HasPrefix(url, "http") {
		return url, nil
	}

	hash := fmt.Sprintf("%x", sha256.Sum256([]byte(url)))
	localPath := filepath.Join(c.BaseDir, hash+".zen")

	if _, err := os.Stat(localPath); err == nil {
		return localPath, nil
	}

	fmt.Printf("Downloading %s...\n", url)
	resp, err := http.Get(url)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return "", fmt.Errorf("failed to download: status %d", resp.StatusCode)
	}

	data, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return "", err
	}

	if err := ioutil.WriteFile(localPath, data, 0644); err != nil {
		return "", err
	}

	return localPath, nil
}

func (c *Cache) GetTranspiled(sourceHash string) (string, bool) {
	localPath := filepath.Join(c.BaseDir, "transpiled", sourceHash+".php")
	if _, err := os.Stat(localPath); err == nil {
		data, err := ioutil.ReadFile(localPath)
		if err == nil {
			return string(data), true
		}
	}
	return "", false
}

func (c *Cache) SaveTranspiled(sourceHash string, phpCode string) error {
	dir := filepath.Join(c.BaseDir, "transpiled")
	if err := os.MkdirAll(dir, 0755); err != nil {
		return err
	}
	localPath := filepath.Join(dir, sourceHash+".php")
	return ioutil.WriteFile(localPath, []byte(phpCode), 0644)
}

File: src/core/cache.rs
use sha2::{Digest, Sha256};
use std::fs;
use std::path::PathBuf;

pub struct Cache {
    pub base_dir: PathBuf,
}

impl Cache {
    pub fn new() -> anyhow::Result<Self> {
        let home =
            dirs::home_dir().ok_or_else(|| anyhow::anyhow!("Could not find home directory"))?;
        let base = home.join(".zenith").join("cache");
        fs::create_dir_all(&base)?;
        Ok(Cache { base_dir: base })
    }

    pub fn get(&self, url: &str) -> anyhow::Result<String> {
        if !url.starts_with("http") {
            return Ok(url.to_string());
        }

        let mut hasher = Sha256::new();
        hasher.update(url.as_bytes());
        let hash = format!("{:x}", hasher.finalize());
        let local_path = self.base_dir.join(format!("{}.zen", hash));

        if local_path.exists() {
            return Ok(local_path.to_string_lossy().to_string());
        }

        println!("Downloading {}...", url);
        let mut response = reqwest::blocking::get(url)?;
        if !response.status().is_success() {
            return Err(anyhow::anyhow!(
                "Failed to download: status {}",
                response.status()
            ));
        }

        let mut out = fs::File::create(&local_path)?;
        response.copy_to(&mut out)?;

        Ok(local_path.to_string_lossy().to_string())
    }

    pub fn get_transpiled(&self, source_hash: &str) -> Option<String> {
        let local_path = self
            .base_dir
            .join("transpiled")
            .join(format!("{}.php", source_hash));
        if local_path.exists() {
            fs::read_to_string(local_path).ok()
        } else {
            None
        }
    }

    pub fn save_transpiled(&self, source_hash: &str, php_code: &str) -> anyhow::Result<()> {
        let dir = self.base_dir.join("transpiled");
        fs::create_dir_all(&dir)?;
        let local_path = dir.join(format!("{}.php", source_hash));
        fs::write(local_path, php_code)?;
        Ok(())
    }

    pub fn get_transpiled_path(&self, source_hash: &str) -> PathBuf {
        self.base_dir
            .join("transpiled")
            .join(format!("{}.php", source_hash))
    }

    pub fn save_runtime(&self, php_code: &str) -> anyhow::Result<PathBuf> {
        let path = self.base_dir.join("runtime.php");
        fs::write(&path, php_code)?;
        Ok(path)
    }
}
